staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go

```diff
@@ -340,62 +299,85 @@ func (c *celAdmissionController) Validate(
 				versionedAttr = va
 			}
 
-			validationResult := bindingInfo.validator.Validate(ctx, versionedAttr, param, celconfig.RuntimeCELCostBudget, authz)
+			var validationResults []ValidateResult
 
-			for i, decision := range validationResult.Decisions {
-				switch decision.Action {
-				case ActionAdmit:
-					if decision.Evaluation == EvalError {
-						celmetrics.Metrics.ObserveAdmissionWithError(ctx, decision.Elapsed, definition.Name, binding.Name, "active")
-					}
-				case ActionDeny:
-					for _, action := range binding.Spec.ValidationActions {
-						switch action {
-						case v1alpha1.Deny:
-							deniedDecisions = append(deniedDecisions, policyDecisionWithMetadata{
-								Definition:     definition,
-								Binding:        binding,
-								PolicyDecision: decision,
-							})
-							celmetrics.Metrics.ObserveRejection(ctx, decision.Elapsed, definition.Name, binding.Name, "active")
-						case v1alpha1.Audit:
-							c.publishValidationFailureAnnotation(binding, i, decision, versionedAttr)
-							celmetrics.Metrics.ObserveAudit(ctx, decision.Elapsed, definition.Name, binding.Name, "active")
-						case v1alpha1.Warn:
-							warning.AddWarning(ctx, "", fmt.Sprintf("Validation failed for ValidatingAdmissionPolicy '%s' with binding '%s': %s", definition.Name, binding.Name, decision.Message))
-							celmetrics.Metrics.ObserveWarn(ctx, decision.Elapsed, definition.Name, binding.Name, "active")
+			if len(params) == 0 {
+				validationResults = append(validationResults, bindingInfo.validator.Validate(ctx, versionedAttr, nil, celconfig.RuntimeCELCostBudget, authz))
+			} else {
```

Comment:
Does this work for the case where the namespaced param selector matches no results and the paramNotFoundAction=ignore?  I was expecting no validation for this case. Does this validate that case with a null param? (I might be missing something)

Reply:
(If this is tested, apologies, I did a quick scan and didn't find it.  If not, can we add a test?)

Reply:
The way it is implemented now is that no params found + Ignore failAction will result in the policy being evaluated with null param.

This does not match the docs or the KEP. I think it must've not survived a refactor. Will add a test for this and correct the logic. Thanks for pointing this out

Reply:
your question got me thinking and I wanted to clarify some of the case logic on the PR and make sure there are tests for all:

1. No paramKind -> Validate with null param
2. ParamKind set on policy, but no paramRef/namespaceParamRef on binding -> skip/pass binding
3. NotFound when using ClusterWideParam -> configuration error
4. NotFound when using NamespaceParamRef w/ Deny Action -> error
5. NotFound when using NamespaceParamRef w/ Allow Action -> skip/pass binding

That is, the only scenario where "no params" means "evaluate with null param" would be if the policy does not set paramKind, correct?

I am also kind of unsure of #2. Previously in this case we've run the policy with `nil` param

Reply:
I think for 2, we will validate with null param?

Reply:
I've fixed the logic so that only in cases 1 and 2 are null params validated

Also ensured we had tests for each case and added where missing

1. TestEmptyParamSource (existing)
2. TestEmptyParamRef (new)
3. TestInvalidParamSourceInstanceName (existing)
4. TestNamespaceParamRefFailActionDeny (new)
5. TestNamespaceParamRefFailActionAllow (new)